### PR TITLE
 Only merge params option if params is a Hash in url_for helper

### PR DESCRIPTION
**Pull request by ghost**
This PR Fixes the problem described in [this original issue](https://github.com/rails/rails/issues/33248) that was fixed [here](https://github.com/rails/rails/pull/33256) and reverted [here](https://github.com/rails/rails/commit/17e4b49292fb8da8a212e8200e2e94c44aedb788)

The fix was reverted because the keyword params is special and is used to make possible to pass parameters that are reserved.

So this fix suggests to only merge params when it is a `Hash`.

The example in the [revert commit](https://github.com/rails/rails/commit/17e4b49292fb8da8a212e8200e2e94c44aedb788) from @rafaelfranca  suggests that params, when used, takes the form of a Hash.

closes #42028
closes #42052

### CODE REVIEW

File: actionpack/lib/action_dispatch/routing/route_set.rb [hunk 1]
```diff
@@ -822,7 +822,7 @@ def url_for(options, route_name = nil, url_strategy = UNKNOWN, method_name = nil
         path = route_with_params.path(method_name)
         params = route_with_params.params
 
-        if options.key? :params
+        if options.key?(:params) && options[:params].is_a?(Hash)
```
Comment (MatheusRich) [inline actionpack/lib/action_dispatch/routing/route_set.rb:825]: Shouldn't we check the second branch first?
```suggestion
        if options[:params].is_a?(Hash) && options.key?(:params)
```

Comment (bquorning) [reply actionpack/lib/action_dispatch/routing/route_set.rb:825]: if `options[:params].is_a?(Hash)` evaluates to true, you know that `options.key?(:params)` will also be true. So the line _could_ be reduced to just `if options[:params].is_a?(Hash)`.

Comment (MatheusRich) [reply actionpack/lib/action_dispatch/routing/route_set.rb:825]: Whoops, I didn't notice the check was `options[:params].is_a?(Hash)` not `options.is_a?(Hash)`.

Comment (chaymaeBZ) [reply actionpack/lib/action_dispatch/routing/route_set.rb]: @bquorning your suggestion is on point 👌🏼 I changed accordingly

**Comment by benoittgt** [comment]
Don't we need a test?

**Comment by sholden** [comment]
@benoittgt I added a test in the pull request @marivaldo just referenced, can pluck it from there.

**Comment by chaymaeBZ** [comment]
@benoittgt I made a test for when params are not a hash. which made me realise that `url_for` is overriding existing params in [this line](https://github.com/chaymaebz/rails/blob/main/actionpack/lib/action_dispatch/routing/route_set.rb#L831).
In other words, if a special param called `params` is being passed, it will get deleted (or replaced). Therefore, in that special case, we want to keep the param called params in `params[:params]`

if you inspect the options that `url_for` built in that test case, you'll get something like 
`{:only_path=>true, :controller=>"c", :action=>"a", :params=>{:params=>"p"}, :path=>"/c/a", :script_name=>"", :user=>nil, :password=>nil}`

**Review by marivaldo — APPROVED** [review]

**Review by lfalcao — APPROVED** [review]

**Comment by zenspider** [comment]
_nudge_

File: actionpack/lib/action_dispatch/routing/route_set.rb [hunk 2]
```diff
@@ -822,8 +822,10 @@ def url_for(options, route_name = nil, url_strategy = UNKNOWN, method_name = nil
         path = route_with_params.path(method_name)
         params = route_with_params.params
 
-        if options.key? :params
+        if options[:params].is_a?(Hash)
```
Comment (casperisfine) [inline actionpack/lib/action_dispatch/routing/route_set.rb:825]: So this broke our app quite extensively as it's not rare for `options[:params]` to be an `ActionController::Parameters`. 

I'll fix it in https://github.com/rails/rails/pull/42603, feel free to chime in.

**Pull request merged**